init/lib/start.go

```diff
@@ -17,24 +17,68 @@ package lib
 import (
 	"io/ioutil"
 	"os"
-	"os/exec"
-	"strconv"
 
 	"github.com/pkg/errors"
+	"gopkg.in/validator.v2"
+	"gopkg.in/yaml.v2"
+
+	"github.com/palantir/go-java-launcher/launchlib"
 )
 
-// StartCommand starts the given command, outputting to var/log/startup.log and writing the resulting process's PID to
-// var/run/service.pid.
-func StartCommand(cmd *exec.Cmd, outputFile *os.File) error {
-	cmd.Stdout = outputFile
-	cmd.Stderr = outputFile
-	if err := cmd.Start(); err != nil {
+func StartService(procCmds []launchlib.ProcCmd) error {
+	for _, procCmd := range procCmds {
+		if err := startCommand(procCmd); err != nil {
+			return errors.Wrap(err, "failed to start at least one process")
+		}
+		if err := writePid(procCmd.Name, procCmd.Cmd.Process.Pid); err != nil {
+			return errors.Wrap(err, "failed to record at least one pid")
+		}
+	}
+	return nil
+}
+
+func startCommand(procCmd launchlib.ProcCmd) (rErr error) {
+	procCmd.Cmd.Stdout = procCmd.Stdout
+	procCmd.Cmd.Stderr = procCmd.Stdout
+	if err := procCmd.Cmd.Start(); err != nil {
 		return errors.Wrap(err, "failed to start command")
 	}
+	return nil
+}
 
-	if err := ioutil.WriteFile(Pidfile, []byte(strconv.Itoa(cmd.Process.Pid)), 0644); err != nil {
+func writePid(name string, pid int) error {
+	var servicePids ServicePids
+	if pidfileExists() {
+		pidfileBytes, err := ioutil.ReadFile(pidfile)
+		if err != nil {
+			return errors.Wrap(err, "failed to read previous pidfile")
+		}
+		if err := yaml.Unmarshal(pidfileBytes, &servicePids); err != nil {
+			return errors.Wrap(err, "failed to deserialize pidfile")
+		}
+		if err := validator.Validate(servicePids); err != nil {
+			return errors.Wrap(err, "failed to deserialize pidfile")
+		}
+	} else {
+		servicePids.PidsByName = make(map[string]int)
+	}
+
+	servicePids.PidsByName[name] = pid
+	servicePidsBytes, err := yaml.Marshal(servicePids)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize pidfile")
+	}
+	if err := ioutil.WriteFile(pidfile, servicePidsBytes, 0666); err != nil {
 		return errors.Wrap(err, "failed to write pidfile")
 	}
 
 	return nil
 }
+
+func pidfileExists() bool {
+	if _, err := os.Stat(pidfile); err != nil {
+		// The only piece of information from the error we care about is if the file exists.
+		return !os.IsNotExist(err)
```

Comment:
Intentionally throwing away a huge class of errors here because we don't care about them here. All this function does is say if the file exists - it isn't designed to notify the caller of anything else. Encountering an error such as "the file does exist but is somehow bogus" is the responsibility of the caller to handle (and is handled).

However, your note about the pidfile disappearing between the invocation of this function and a possible subsequent read call is a good flag, and is sufficient reason to not look before leaping.